nighthawk/test/server/http_test_server_filter_integration_test.cc

```diff
@@ -0,0 +1,262 @@
+        "tcp://{}:{}", Envoy::Network::Test::getLoopbackAddressUrlString(ip_version), port));
+    return makeSingleRequest(addr, method, url, body, type, host, content_type,
+                             request_header_delegate);
+  }
+
+  Envoy::BufferingStreamDecoderPtr
+  makeSingleRequest(const Envoy::Network::Address::InstanceConstSharedPtr& addr,
+                    const std::string& method, const std::string& url, const std::string& body,
+                    Envoy::Http::CodecClient::Type type, const std::string& host,
+                    const std::string& content_type,
+                    std::function<void(Envoy::Http::HeaderMapImpl&)> request_header_delegate) {
+
+    testing::NiceMock<Envoy::Stats::MockIsolatedStatsStore> mock_stats_store;
+    Envoy::Event::GlobalTimeSystem time_system;
+    Envoy::Api::Impl api(Envoy::Thread::threadFactoryForTest(), mock_stats_store, time_system,
+                         Envoy::Filesystem::fileSystemForTest());
+    Envoy::Event::DispatcherPtr dispatcher(api.allocateDispatcher());
+    std::shared_ptr<Envoy::Upstream::MockClusterInfo> cluster{
+        new testing::NiceMock<Envoy::Upstream::MockClusterInfo>()};
+    Envoy::Upstream::HostDescriptionConstSharedPtr host_description{
+        Envoy::Upstream::makeTestHostDescription(cluster, "tcp://127.0.0.1:80")};
+    Envoy::Http::CodecClientProd client(
+        type,
+        dispatcher->createClientConnection(addr, Envoy::Network::Address::InstanceConstSharedPtr(),
+                                           Envoy::Network::Test::createRawBufferSocket(), nullptr),
+        host_description, *dispatcher);
+    Envoy::BufferingStreamDecoderPtr response(new Envoy::BufferingStreamDecoder([&]() -> void {
+      client.close();
+      dispatcher->exit();
+    }));
+    Envoy::Http::StreamEncoder& encoder = client.newStream(*response);
+    encoder.getStream().addCallbacks(*response);
+
+    Envoy::Http::HeaderMapImpl headers;
+    headers.insertMethod().value(method);
+    headers.insertPath().value(url);
+    headers.insertHost().value(host);
+    headers.insertScheme().value(Envoy::Http::Headers::get().SchemeValues.Http);
+    if (!content_type.empty()) {
+      headers.insertContentType().value(content_type);
+    }
+    request_header_delegate(headers);
+    encoder.encodeHeaders(headers, body.empty());
+    if (!body.empty()) {
+      Envoy::Buffer::OwnedImpl body_buffer(body);
+      encoder.encodeData(body_buffer, true);
+    }
+
+    dispatcher->run(Envoy::Event::Dispatcher::RunType::Block);
+    return response;
+  }
+
+  void testWithResponseSize(int response_size, bool expect_header = true) {
+    Envoy::BufferingStreamDecoderPtr response = makeSingleRequest(
+        lookupPort("http"), "GET", "/", "", downstream_protocol_, version_, "foo.com", "",
+        [response_size](Envoy::Http::HeaderMapImpl& request_headers) {
+          const std::string header_config = fmt::format("{{response_size:{}}}", response_size);
```

Comment:
This is fine for such a simple message, if it gets more complicated I'd suggest building up the proto and then converting it to JSON.